src/ss.rs

```diff
@@ -0,0 +1,330 @@
+use rand::Rng;
+use sha3::{Digest, Sha3_256};
+
+pub struct BitXORShare {}
+
+impl BitXORShare {
+    #[must_use]
+    pub fn create(value: u32, share_amount: u8) -> Vec<u32> {
+        if share_amount < 2 {
+            panic!("Amount of shares must be at least 2.");
+        }
+        let mut rng = rand::thread_rng();
+        let mut last_sh = value;
+        let mut sh_vec = vec![];
+        for _i in 0..share_amount - 1 {
+            let sh = rng.gen::<u32>();
+            sh_vec.push(sh);
+            last_sh ^= sh;
+        }
+        sh_vec.push(last_sh);
+        sh_vec
+    }
+
+    pub fn combine(shares: Vec<u32>) -> u32 {
+        let mut value = 0;
+        for sh in shares {
+            if value == 0 {
+                value = sh;
+            } else {
+                value ^= sh;
+            }
+        }
+        value
+    }
+}
+
+pub struct SecureCoinTossing {}
+
+impl SecureCoinTossing {
+    pub fn get_random_and_hash() -> (u128, String) {
+        // generate a u128 random integer.
+        let mut rng = rand::thread_rng();
+        let r = rng.gen::<u128>();
+
+        // generate hash for the random integer's bit string.
+        let mut hasher = Sha3_256::new();
+        hasher.update(format!("{:b}", r));
+        let h: String = format!("{:X}", hasher.finalize());
+        (r, h)
+    }
+
+    pub fn get_random() -> u128 {
+        // generate a u128 random integer.
+        let mut rng = rand::thread_rng();
+        rng.gen::<u128>()
+    }
+
+    pub fn verify_random_and_hash(r: u128, h: String) -> bool {
+        // generate hash for the random integer's bit string.
+        let mut hasher = Sha3_256::new();
+        hasher.update(format!("{:b}", r));
+        let hv: String = format!("{:X}", hasher.finalize());
+        h == hv
+    }
+
+    pub fn calculate_shared_random(r1: u128, r2: u128) -> u128 {
+        r1 ^ r2
+    }
+}
+
+pub struct RejectionSampling {}
+
+impl RejectionSampling {
+    pub fn generate_permutation(bit_string_as_int: u128, dimension: usize) -> Vec<usize> {
+        // Check if input dimension can be supported.
+        // bit_string_as_int is 128 bit. It only enough to generate 4 bit dimension.
+        // E.g. if demension is 30, we needs 5 bits to generate 1 integer number
+        // between 1 to 30, and we need at least 30 group of 5 bits to generate
+        // 20 number to meet the demension. 5 bits times 30 is bigger 128 bit.
+        let max_dimension = 2usize.pow(4);
+        if dimension > max_dimension {
+            panic!(
+                "dimension ({}) > {} is not supported!",
+                dimension, max_dimension
+            );
+        }
+
+        // calculate how many bit we need to generate an integer number in dimension.
+        // ceiling of log2(n), n is the number of dimensions.
+        let bit = ((dimension) as f64).log2().ceil() as u8;
```

Comment:
This calculation is problematic as the use of floating point leads to inaccuracies.  It's also slower than it needs to be.  You probably want to look at u128::leading_zeros (and size_of) for this.